vllm/model_executor/models/molmo.py

```diff
@@ -0,0 +1,1130 @@
+                kv_caches[i - self.start_layer],
+                attn_metadata,
+                residual,
+            )
+        if not get_pp_group().is_last_rank:
+            return IntermediateTensors({
+                "hidden_states": hidden_states,
+                "residual": residual
+            })
+        if residual is not None:
+            hidden_states, _ = self.norm(hidden_states, residual)
+        else:
+            hidden_states = self.norm(hidden_states)
+        return hidden_states
+
+
+cached_get_processor = lru_cache(get_processor)
+
+
+def get_num_patches(num_tiles, crop_patches, left_margin, right_margin, pooling_size):
+    crop_window_patches = crop_patches - (left_margin + right_margin)
+    if num_tiles > 1:
+        left_crop_window_patches = (crop_window_patches + left_margin + pooling_size - 1) // pooling_size * pooling_size
+        middle_crop_window_patches = (crop_window_patches + pooling_size - 1) // pooling_size * pooling_size
+        right_crop_window_patches = (crop_window_patches + right_margin + pooling_size - 1) // pooling_size * pooling_size
+        return left_crop_window_patches + (num_tiles - 2) * middle_crop_window_patches + right_crop_window_patches
+    else:
+        single_crop_window_patches = (crop_patches + pooling_size - 1) // pooling_size * pooling_size
+        return single_crop_window_patches
+
+
+def get_tokens(tiling_h, tiling_w, crop_patches, left_margin, right_margin, pooling_size):
+    h = get_num_patches(tiling_h, crop_patches, left_margin, right_margin, pooling_size)
+    w = get_num_patches(tiling_w, crop_patches, left_margin, right_margin, pooling_size)
+    per_row = w // pooling_size + 1
+    joint = per_row * (h // pooling_size) + 2
+    image_token_length = (crop_patches + pooling_size - 1) // pooling_size
+    resize = (image_token_length + 1) * image_token_length + 2
+    return resize + joint
+
+
+def get_max_tokens(max_crops, crop_patches, left_margin, right_margin, pooling_size):
+    tilings = []
+    for i in range(1, max_crops+1):
+        for j in range(1, max_crops + 1):
+            if i * j <= max_crops:
+                tilings.append((i, j))
+    tokens = [get_tokens(tilings[i][0], tilings[i][1], crop_patches, left_margin, right_margin, pooling_size) for i in range(len(tilings))]
+    return max(tokens)
+
+
+def get_max_molmo_image_tokens(ctx: InputContext) -> int:
+    processor = cached_get_processor(ctx.model_config.model, trust_remote_code=True)
+    image_processor = processor.image_processor
+    max_llm_image_tokens = get_max_tokens(
+        image_processor.max_crops,
+        image_processor.base_image_input_size[0] // image_processor.image_patch_size,
+        image_processor.overlap_margins[0],
+        image_processor.overlap_margins[1],
+        2,
+    )
+    return max_llm_image_tokens
+
+
+def image_input_mapper_for_molmo(
+    ctx: InputContext,
+    data: Union[Image.Image, List[Image.Image]],
+):
+    return MultiModalInputs(data)
```

Comment:
Can you update this regarding my comment in the `offline_inference_molmo.py`?

Reply:
Could you explain in a bit more detail?

Reply:
I think this is outdated, it was about accepting PIL images into this function rather than numpy arrays. Since this function works with PIL no additional changes are needed.